### PR TITLE
Remove xsltproc dependency

### PR DESCRIPTION
**Pull request by isbm**
Changes proposed in this pull request:
* Do not call 3rd party XPath processor (`xsltproc`) but use Python library instead
* Allow XMLDescription to be loaded also from the XML in the string, instead of just a physical file

### CODE REVIEW

File: kiwi/xml_description.py [hunk 1]
```diff
@@ -52,9 +53,9 @@ class XMLDescription(object):
         path to base XML description file
 
     """
-    def __init__(self, description, derived_from=None):
+    def __init__(self, description=None, derived_from=None, content=None):
         self.description_xslt_processed = NamedTemporaryFile()
-        self.description = description
+        self.description = content and BytesIO(content) or description
```
Comment (aplanas) [inline kiwi/xml_description.py]: `self.description = BytesIO(content) if content else description`

But I found this a bit wrong that content have move priority than description when both are passed. I would drop content here

File: kiwi/xml_description.py [hunk 2]
```diff
@@ -105,11 +106,13 @@ def __parse(self):
...
-    def __xsltproc(self):
-        Command.run(
-            [
-                'xsltproc', '-o', self.description_xslt_processed.name,
-                Defaults.get_xsl_stylesheet_file(),
-                self.description
-            ]
-        )
+    def _xsltproc(self):
+        '''
```
Comment (aplanas) [inline kiwi/xml_description.py]: """?

File: kiwi/xml_description.py (hunk 1 above)
Comment (isbm) [inline kiwi/xml_description.py]: You're saying exactly the same thing, just a different syntax. :wink:

Comment (aplanas) [inline kiwi/xml_description.py]: Ah sure, is a matter of clarity. `and` and `or` was idiomatic like in 2.4 (10 years ago?) Still confusing the precedence of content over description for the value of self.description. This precedence is even more clear with the `if` expression

Comment (isbm) [inline kiwi/xml_description.py]: Matter of taste.

File: kiwi/xml_description.py (hunk 2 above)
Comment (isbm) [inline kiwi/xml_description.py]: Actually, having developing Salt, we hate double-quotes there and using single all everywhere, although PEP8 leans to the double-quotes. Old habits die hard...

While I fixed this, I would suggest not to make it religious, unless @schaefi has a strict PyLint on this.

File: kiwi/xml_description.py [hunk 3]
```diff
@@ -52,9 +54,9 @@ class XMLDescription(object):
         path to base XML description file
 
     """
-    def __init__(self, description, derived_from=None):
+    def __init__(self, description=None, derived_from=None, content=None):
         self.description_xslt_processed = NamedTemporaryFile()
-        self.description = description
+        self.description = content and BytesIO(bytes(content)) or description
```
Comment (aplanas) [inline kiwi/xml_description.py]: What do you want to accomplish here? I see three problems

* hard to read because Boolean circuits instead of using Python
* `content` over `description` (why can't you simply not use description instead of content? or if content is a better name, why not to refactor and change the name of description to content?)
* type mangling, why force it to byte instead of using string? Some python 3 issue? Why not to fix the type issue in the caller and not in the constructor?

Comment (isbm) [inline kiwi/xml_description.py]: > What do you want to accomplish here?

What does it do, you think?

Comment (aplanas) [inline kiwi/xml_description.py]: @isbm I am sorry if I am upsetting you, not my intention.

Code is to be read so others can understand it. You are submitting the code, and I am this 'other' that is trying to understand it, and to do that I am making some reasonable and honest technical questions. You can decide to answer it or not, but in any case, please, avoid this kind of bullshit.

Comment (isbm) [inline kiwi/xml_description.py]: > hard to read
That's [normal](http://gamedevwithoutacause.com/?p=1329). Also could you please watch your tone? If you don't like it, PR please. Bye.

Comment (schaefi) [inline kiwi/xml_description.py]: I only see a problem with the use of self.description. The scope of it is now two fold, it could be either the path to a file name or raw content which is not a problem for etree.parse but at other places e.g if an exception occurs we use self.description in the exception message and it would make only sense if it would be in the scope of a filename. In addition the member variable parse.description_dir should be None if raw content will be used and would be set to the content which might be confusing

How about dealing with self.description and self.content instead of merging it at the object construction level
e.g

```python
self.description = description
self.content = content

...

etree.parse(self.content and BytesIO(bytes(self.content)) or self.description)

...handle output of KiwiDescriptionInvalid exception accordingly

```

I think that would put the logic at the place where it is needed and follows more the explicit is better than implicit python advise

Thoughts ?

Comment (isbm) [inline kiwi/xml_description.py]: OK, the exception handling issue makes perfect sense as well as the directory variable. Will change that accordingly.

Comment (aplanas) [inline kiwi/xml_description.py]: @schaefi Aaah more clear IMHO. There are some bits that can help too:

* Raise an exception when both description and content are provided. The prevalence of one over another at least needs to be documented in the header (also missing)
* I am a bit worried with the type mangling. `content` now expect bytes and can misbehave in unicode. The use of `six` indicates me that is expected to work on Py2 and Py3 so this can be a problem. Can be helpful to add test cases for unicode and str (py2) and str (py3)

I'm sure that @isbm had good reasons to use bytes (I asked to understand them), but I am curious why not use `ET.fromstring` or `ET.parse` directly to manage file and string cases.

One note also, `from builtins import bytes` is not going to work in py2, but maybe this test is only executed in Py3

Comment (isbm) [inline kiwi/xml_description.py]: @aplanas tests are passing on both sides. In Salt we use `ext` from the beginning. Yet we have way less noise in the comments. :wink:

**Comment by schaefi** [comment]
@isbm do you plan to work on this request next week ? if not it's ok and if ok with you I would take over the branch because it's a great idea to get rid of the external xsltproc dependency for the next release which is scheduled for next week. Just let me know, thanks

**Comment by isbm** [comment]
@schaefi Yes, in general I plan. But if you have a time, you could also push changes to this branch, of course!

File: kiwi/xml_description.py [hunk 4]
```diff
@@ -86,17 +95,19 @@ def load(self):
                 '%s: %s' % (type(e).__name__, format(e))
             )
         if not validation_ok:
-            raise KiwiDescriptionInvalid(
-                'Schema validation for %s failed' % self.description
-            )
+            if self.description:
+                message = 'Schema validation for %s failed' % self.description
```
Comment (isbm) [inline kiwi/xml_description.py]: Nitpick of formatting: lately we use in Python `.format(...)` instead of this way. Example:
```python
message = 'Schema validation of {0} had failed'.format(self.description)
```
Or with parameters:
```python
message = 'Schema validation of {descr} had failed'.format(descr=self.description)
```
JFYI.

File: test/unit/xml_description_test.py [hunk 5]
```diff
@@ -27,35 +47,49 @@ def test_load_schema_import_error(self, mock_xslt, mock_relax):
...
+        self.description_from_file.load()
 
     @raises(KiwiValidationError)
     @patch('lxml.etree.RelaxNG')
     @patch('lxml.etree.parse')
     @patch.object(XMLDescription, '_xsltproc')
-    def test_load_schema_validation_error(
+    def test_load_schema_validation_error_from_file(
         self, mock_xslt, mock_parse, mock_relax
     ):
```
Comment (isbm) [inline test/unit/xml_description_test.py]: This is compilable but I would suggest not to format it that way due to [indentation as per PEP8](https://www.python.org/dev/peps/pep-0008/#indentation). I would offer several ways.

As a member of Emacs religion:
```python
    def test_load_schema_validation_error_from_file(self, mock_xslt,
                                                    mock_parse, mock_relax):
    # Code here
```
Or:
```python
def test_load_schema_validation_error_from_file(
        self, mock_xslt, mock_parse, mock_relax):
    # Code here
```
Or:
```python
def test_load_schema_validation_error_from_file(self,
                                                mock_xslt,
                                                mock_parse,
                                                mock_relax):
    # Code here
```
In all cases the code starts from the left side from `):` thing.

File: kiwi/exceptions.py [hunk 6]
```diff
@@ -131,6 +131,10 @@ class KiwiDataStructureError(KiwiError):
     pass
 
 
+class KiwiDescriptionConflict(KiwiError):
+    pass
```
Comment (isbm) [inline kiwi/exceptions.py]: Instead of `pass` I would just add a documentation here. Something like:
```python
# No
class KiwiDescriptionConflict(KiwiError):
    pass

# Yes
class KiwiDescriptionConflict(KiwiError):
    """
    Documentation, that describes a purpose of this class.
    Below of this documentation string there is nothing else.
    """
```

File: kiwi/xml_description.py [hunk 7]
```diff
@@ -52,12 +53,20 @@ class XMLDescription(object):
     * :attr:`derived_from`
         path to base XML description file
 
+    * :attr:`xml_content`
+        XML description data as content string
     """
-    def __init__(self, description=None, derived_from=None, content=None):
+    def __init__(self, description=None, derived_from=None, xml_content=None):
```
Comment (isbm) [inline kiwi/xml_description.py]: Well, if we are renaming `content` to `xml_content` (are we expect something like `json_content` or `yaml_content` too?) then the `description` is actually just a path to the file. Maybe it can be `path` now?

I would keep `content` instead.

File: kiwi/xml_description.py [hunk 8]
```diff
@@ -108,10 +119,21 @@ def __parse(self):
...
         """
-        transform = etree.XSLT(etree.parse(Defaults.get_xsl_stylesheet_file()))
+        xslt_transform = etree.XSLT(
+            etree.parse(Defaults.get_xsl_stylesheet_file())
+        )
+
+        xml_source = self.description
+        if self.xml_content:
+            xml_source = BytesIO(bytes(self.xml_content))
+
```
Comment (isbm) [inline kiwi/xml_description.py]: I always try to avoid re-assigning and thus I would write it:
```python
if self.xml_content:
    xml_source = BytesIO(bytes(self.xml_content))
else:
    xml_source = self.description
```
But this is looks like a Pascal in my student's ages. Therefore use ternary operator:
```python
xml_source = BytesIO(bytes(self.xml_content)) if self.xml_content else self.description
```
Or 1 char shorter:
```python
xml_source = self.xml_content and BytesIO(bytes(self.xml_content)) or self.description
```

File: test/unit/xml_description_test.py [hunk 9]
```diff
@@ -3,22 +3,42 @@
 
 from .test_helper import *
 
-from kiwi.exceptions import (
-    KiwiSchemaImportError,
-    KiwiValidationError,
-    KiwiDescriptionInvalid,
-    KiwiDataStructureError
-)
+from kiwi.exceptions import *
```
Comment (isbm) [inline test/unit/xml_description_test.py]: Please don't. Stars are _very bad_ style and [PEP8 hates that](https://www.python.org/dev/peps/pep-0008/#imports).

Comment (schaefi) [inline test/unit/xml_description_test.py]: we "allow" the wildcard import only in the tests, wouldn't that be fair enough ?
exceptions not handled in a test will result in a coverage error, thus I thought it is ok to have a wildcard import for exceptions in tests

File: kiwi/exceptions.py (hunk 6 above)
Comment (schaefi) [inline kiwi/exceptions.py]: I agree, I will add doc strings explaining the exceptions in an extra pull request ok ?

File: test/unit/xml_description_test.py (hunk 9 above)
Comment (isbm) [inline test/unit/xml_description_test.py]: Test is also code. :wink: Personally I don't do it anywhere and I always want to know what is imported. But your call...

File: kiwi/xml_description.py (hunk 4 above)
Comment (schaefi) [inline kiwi/xml_description.py]: nice, that is easier to read, will make use of it

File: kiwi/exceptions.py (hunk 6 above)
Comment (schaefi) [inline kiwi/exceptions.py]: opened branch exceptions-doc-strings to track this one

**Comment by schaefi** [comment]
if you think we are good to go, give me a thumbs up

I'll fix the doc strings for exceptions next... boring but must be done :smile:

File: test/unit/xml_description_test.py [hunk 10]
```diff
@@ -1,67 +1,109 @@
...
+        self.description_from_file = XMLDescription(description='description')
+        self.description_from_data = XMLDescription(xml_content=test_xml)
+
+    @raises(KiwiDescriptionConflict)
+    def test_constructor_conflict(self):
+        XMLDescription(description='description', xml_content='content')
+
+    def test_load_schema_from_xml_content(self):
+        parsed = self.description_from_data.load()
+        assert parsed.get_schemaversion() == '6.3'
```
Comment (isbm) [inline test/unit/xml_description_test.py]: Maybe the schema version can be picked up from somewhere, so we don't need to fix this test all the time XSL template changes? (Just a thought)

Comment (schaefi) [inline test/unit/xml_description_test.py]: good idea

File: kiwi/exceptions.py (hunk 6 above)
Comment (schaefi) [inline kiwi/exceptions.py]: request #97 open for review

**Comment by isbm** [comment]
Perfect. Go for it. :+1:
